packages/byods/src/BYODS.ts

```diff
@@ -1,47 +1,389 @@
+    try {
+      const response = await fetch(`${this.baseUrl}/applications/${this.serviceAppId}/token`, {
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${personalAccessToken}`,
+          'Content-Type': 'application/json',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          targetOrgId: orgId,
+          clientId: this.config.clientId,
+          clientSecret: this.config.clientSecret,
+        }),
+      });
+
+      if (!response.ok) {
+        throw new Error(`Failed to create service app authorization: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+
+      return data;
+    } catch (error) {
+      console.error('Error creating service app authorization:', error);
+      throw error;
     }
+  }
+
+  /**
+   * Retrieve a new service app token after authorization.
+   * @param {string} orgId - The organization ID.
+   * @param {string} token - The service app owner's personal access token or token from an integration that has the scope `spark:applications_token`.
+   * @returns {Promise<void>}
+   */
+  public async retrieveTokenAfterAuthorization(orgId: string, token: string): Promise<void> {
+    try {
+      const response = await fetch(`${this.baseUrl}/applications/${this.serviceAppId}/token`, {
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${token}`,
+          'Content-Type': 'application/x-www-form-urlencoded',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          targetOrgId: orgId,
+          clientId: this.config.clientId,
+          clientSecret: this.config.clientSecret,
+        }),
+      });
 
-    // Use this.token.access_token to make authenticated requests
-    const response = await fetch(endpoint, {
+      if (!response.ok) {
+        throw new Error(`Failed to retrieve token: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+    } catch (error) {
+      console.error('Error retrieving token after authorization:', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Get the service app authorization for a given organization ID.
+   * @param {string} orgId - The organization ID.
+   * @returns {Promise<ServiceAppAuthorization>}
+   */
+  public async getServiceAppAuthorization(orgId: string): Promise<ServiceAppAuthorization> {
+    if (!this.serviceAppAuthorizations[orgId]) {
+      return Promise.reject(new Error('Service app authorization not found'));
+    }
+
+    return Promise.resolve(this.serviceAppAuthorizations[orgId]);
+  }
+
+  /**
+   * Refresh the access token using the refresh token.
+   * @param {string} orgId - The organization ID.
+   * @returns {Promise<void>}
+   */
+  public async refreshServiceAppAccessToken(orgId: string): Promise<void> {
```

Comment:
I see mix of functions here: methods for service app-specific functionalities and CRUD API functions. Are we planning to create a hierarchy in there will be different files and classes to handle different functionalities and base class should just be used as the starting point?

Reply:
I've split the class into different classes for better organization now